### ISSUE TITLE
Harmful paging in Percolator results

### ISSUE THREAD

**Issue by dennisgorelik** [elastic/elasticsearch#25308]
When [percolator index structure was "standardized"](https://www.elastic.co/blog/elasticsearch-percolator-continues-to-evolve) - it gained unwanted features from standard search index:
1)	Page size limit.
2)	Match score and other unneeded information.

There are no practical advantages of having these misfeatures, but there are significant problems associated with them.
In particular, from business perspective, limiting page size on percolator results just does not make sense: if I have 1 million subscribers (percolator queries) to a specific keyword in a document, why would I want to receive matches only to a small subset (one page) of these subscribers? Quite the contrary: from the business perspective I would always want to receive matches to all percolator subscribers.
Why would I want to rank and order my subscribers when I output results? I simply want to notify all my subscribers about new document that is matching their percolator queries.

Current "standardized index" implementation of percolator index – makes it very easy to make a mistake by outputting only a small subset of matching percolator queries. This subset size is only 10 (sic!) by default.
But if I want to specify higher page size – it can NOT be more than 10,000 without index redesign.
Such implementation is clearly counterproductive to the percolator business logic.

So why had ElasticSearch team decided to break backward compatibility and merge Percolator into a standard search index format?

This is the excuse:

> https://www.elastic.co/blog/elasticsearch-percolator-continues-to-evolve
> Prior to 5.0, all percolator queries need to be executed on this in-memory index in order to verify whether the query matches. So the idea is that the less queries that need to be verified by the in-memory index the faster the percolator executes.

In my first reading of that ambiguous claim I thought that ElasticSearch would be able to automatically detect what percolator queries is ok to skip, so it would, effectively, improve percolator performance.
However the correct interpretation actually is that application developer in ElasticSearch 5.4 has an option to tag percolator queries (alerts), and then write code that would help percolator to skip alerts that have no chance to being triggered by a document we percolate.
But the problem is that it is very hard to come up with such "alerts skipping" algorithm. Percolator is so valuable in the first place exactly because of that ability to determine what alerts match and what alerts do not!

In addition to getting these bugs and misfeatures, with ES 5.4 we got [3 times slower percolator](https://github.com/elastic/elasticsearch/issues/25278).

I do not see any practical positive result in squeezing percolator structure into the standard ElasticSearch index structure. Main search and percolator search are simply different features that should be treated differently.

**Comment by martijnvg** [comment]
> In particular, from business perspective, limiting page size on percolator results just does not make sense: if I have 1 million subscribers (percolator queries) to a specific keyword in a document, why would I want to receive matches only to a small subset (one page) of these subscribers? Quite the contrary: from the business perspective I would always want to receive matches to all percolator subscribers

Returning 1M results (either via the search api or the old percolate api) is putting a lot of pressure on an ES cluster and can cause nodes to fall over. That it was previously allowed was a design flaw. Instead you should paginate using [search after](https://www.elastic.co/guide/en/elasticsearch/reference/5.4/search-request-search-after.html) or a [scrolled search](https://www.elastic.co/guide/en/elasticsearch/reference/5.4/search-request-scroll.html). 

> Why would I want to rank and order my subscribers when I output results? I simply want to notify all my subscribers about new document that is matching their percolator queries.

There are use cases where ranking is wanted. For example news monitor queries containing sloppy phrase queries, some queries are a better match because matches occurred closer to each other. 

If you're not interested in ranking you can easily turn it off, by wrapping the `percolate` query in a `constant_score` query. 

> However the correct interpretation actually is that application developer in ElasticSearch 5.4 has an option to tag percolator queries (alerts), and then write code that would help percolator to skip alerts that have no chance to being triggered by a document we percolate.

The percolator tries to tag the queries automatically based on the containing query terms. However it can't do this for all percolator queries, because the percolator doesn't know how to extract meaningful information during indexing for all queries. This is a work in progress and will get better over time. It already has shown a significant performance improvement for cases where the percolator was able to analyze the percolator query correctly at index time.

**Comment by dennisgorelik** [comment]
### Preventing nodes from falling over
@martijnvg 
Thank you – I see your point. I understand the importance of limiting batch size for performance and stability reasons.

Recently we (at postjobfree.com) reworked a lot of our SQL Server batch queries to limit their size in order to make server more stable and reduce number of timeouts.
But most of the time, the batch size limit we put - we applied NOT to the output of these batch queries, but to the number of raw records that we process in a single batch.
Usually we put limit on number of raw records we process.
Sometimes we put limit on the number of records in the second sub-step of the batch processing, because performance impact of the first sub-step was very small due to a suitable index on a table.
But putting limit on the final output - usually does not protect us from overstressing the server, because the output of that query could be very small in size or even nonexistent and not proportional to the amount of underlying work at all. 

Similarly, in ElasticSearch percolator, if we limit batch size of “hard-to-process” records for performance reasons, then the most important part is to limit the number of records we process (number of potentially matching alerts), not the output size (number of matches percolate query can output).
The reason for that choice of size limit is that the process to detect matches in percolator is much more resource hungry than the process of outputting already found matches.

If you split resource consumption by percolate query into 2 buckets: finding matches vs outputting results – what is the typical proportion between them? My guess that the proportion would be something like 100:1 (percolation itself would consume about 100 times more server resources than generating and sending output would consume).

My main point here is that percolator queries are still different enough from regular search queries and therefore should be treated differently:
- Different mechanism of applying batch size limit (limit on number of alerts instead of limit on output size).
- Different default page size (default of 10 output records just does not make sense in percolator context).
- Different defaults of what to include into output (by default percolator queries should be unranked and naked – matching alerts IDs only and possibly document ID in case of mpercolate).
Making percolate queries even heavier by wrapping them in a constant_score query – in order to turn off ranking – is a weird design choice. Normally, the more output results you want to get – the fluffier your query would be.

**Comment by martijnvg** [comment]
On each shard where individual results are being collected, a sorted list needs to be maintained that is equal to the `from` + `size` you have specified in the percolate / search request. Each shard sends this list (Lucene docid, sort value tuple) to node that coordinated the request The node that coordinates the search / percolate operation also needs to merge N shards results to return a top N to the client. So significantly increasing the number of hits to be returned has a significant overhead.

> My main point here is that percolator queries are still different enough from regular search queries and therefore should be treated differently:

I disagree with you here. For the technical reason explained above, having a different default page size or batch size limit is a bad idea and can lead to performance problems (maybe not initially, but when many requests are executed in parallel). This applies for the old percolator too (the infrastructure it uses is different, but the collecting queries, merging shard level results is the same).

Search after is a perfect mechanism in your case to fetch all the matching alerts.

I think that whether percolator queries should be ranked by default depends on the use case. Since the percolator now uses the search infrastructure, scores will be computed by default. I don't think that is a big issue, since that can easily be turned off in case it isn't desired.

On the bright side by reusing the search infra structure for the percolator, a lot of requested percolator features where immediately implemented. There were many requests for including more than just the id in the response (query metadata which was part of the percolator query's source), sorting percolator queries by metadata fields and having some form of positional scoring to name a few. The percolator having its own infrastructure caused these features never to be implemented in the old percolator, because we would have to duplicate a lot of logic. The percolator infrastructure already contained a lot of duplicated logic. Also there have been many occasions where bugs where fixed in search api and took a while to be fixed in the percolator api.

> Making percolate queries even heavier by wrapping them in a constant_score query

This doesn't make the query heavier. Whatever the this query wraps will completely skip scoring, the `constant_score` query isn't doing any additional operations by itself.

**Comment by dennisgorelik** [comment]
> a sorted list needs to be maintained that is equal to the from + size you have specified in the percolate / search request

This is argument _against_ implementing standard ["from + size" paging](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-from-size.html) for percolator, because getting the tail of the results would stress out ElasticSearch cluster.

> So significantly increasing the number of hits to be returned has a significant overhead.

Could you please quantify "significantly increasing the number"?
For example, what would be faster:
a) Make 1000 requests with 10 resulting alerts each (default page size);
b) Make 1 request with 10000 resulting alerts;
?

> performance problems ... when many requests are executed in parallel

Why would there be many percolator requests executed in parallel?
Percolator is usually executed not from a front-end, but from the backend (where number of running threads is carefully controlled).

> Search after is a perfect mechanism in your case to fetch all the matching alerts.

Not really.
While "search after" seems to be an inevitable option in ElasticSearch 5.4, there is a much better option:
1) Do NOT use ElasticSearch 5.4
2) Do NOT use standard ElasticSearch for percolation at all and use [Elastic Batch Percolator plugin on top of ElasticSearch 1.7](https://github.com/meltwater/elasticsearch-batch-percolator).
We measured batch percolator performance on our 220,000 alerts, and the performance seem to be about 50 times faster than percolator performance on ElasticSearch 1.6.
That is 150 times faster than percolator performance on ElasticSearch 5.4.

That huge performance gain seems surreal, but that is what possible if performance is the goal and the bloat is removed.

> Since the percolator now uses the search infrastructure, scores will be computed by default

That means adding bloat by default.

> a lot of requested percolator features where immediately implemented

Were these requested features really needed?
Are they actually getting used and make ElasticSearch percolator more useful and easier to use?

I remember that once I requested a feature (on a totally different product) and actually got that feature implemented (because it was requested by many users like me). We never used that feature and I am pretty sure other users did not use it too, because it did not add meaningful value, but still required an effort to implement.
That "lot of requested percolator features" may actually fit into the same category: seems to be nice to have and good in theory, but in practice - not really that important.
The most important feature in ElasticSearch percolator is the speed of execution. The second most important - ease of use.
ElasticSearch 5.4 delivered a loss in both these dimensions.

> The percolator infrastructure already contained a lot of duplicated logic. 

So do code reuse when functionality is naturally identical.
But do not force reuse of functionality that is not needed.
Not only such forced reuse hurt the percolator, it can hurt the main ElasticSearch too, because the bloat that would come from marrying Percolator and Main search can make main ElasticSearch less efficient.

> This doesn't make the query heavier.

At a minimum, extra wrapper makes work of query parser slower.
Of course if the core of percolator can barely crawl then query parsing time is not that important.
But in Elastic Batch Percolator we noticed 6% increase in speed (from ~5.5 seconds per down to ~5.2 seconds per 1000 percolated alerts) when we reduced query size by ~11% (from "bool/must" queries to "filtered" queries that ES 5.4 no longer supports)).

Example of fluffier and slower "bool/must" query:
```
{
        "query": {
                "bool": {
                        "must": [
                        {
                                "query_string": {
                                        "fields": ["JobTitle",
                                        "JobDescription",
                                        "CompanyName",
                                        "Salary"],
                                        "query": "\"UNIQUESALARY77777\"",
                                        "analyzer": "pjfqueryanalyzer"
                                }
                        },
                        {
                                "filtered" : {
                                        "query" : {
                                                "match_all" : {}
                                        },
                                        "filter" : {
                                                "geo_distance" : {
                                                        "distance" : "10mi",
                                                        "Location" : {
                                                                "lat":48.856614,
                                                                "lon":2.3522219
                                                        }
                                                }
                                        }
                                }
                        }]
                }
        },
        "SystemUpdateDate": 1497630841197
}
```
Example of faster (and not longer supported in ES 5.4) query:
```
{
        "query": {
                "filtered": {
                        "query": {
                                "query_string": {
                                        "fields": ["JobTitle",
                                        "JobDescription",
                                        "CompanyName",
                                        "Salary"],
                                        "query": "\"UNIQUESALARY77777\"",
                                        "analyzer" : "pjfqueryanalyzer"
                                }
                        },
                        "filter" : {
                                "and" : [{
                                        "geo_distance" : {
                                                "distance" : "10mi",
                                                "Location" : {
                                                        "lat":48.856614,
                                                        "lon":2.3522219
                                                }
                                        }
                                }]
                        }
                }
        },
        "SystemUpdateDate": 1498105034118
}
```

**Comment by martijnvg** [comment]
I think the main issue for the performance loss you report here, is that preselecting of percolator queries isn't excluding that much percolator queries. Like I said in #25278 a cache (but a bit different than before) should maybe be added. If that would be added then I expect that performance would be similar to what you experienced in ES 1.6. (effectively caching the query parsing part)

> This is argument against implementing standard "from + size" paging for percolator, because getting the tail of the results would stress out ElasticSearch cluster.

Yes.

> Could you please quantify "significantly increasing the number"?
For example, what would be faster:
a) Make 1000 requests with 10 resulting alerts each (default page size);
b) Make 1 request with 10000 resulting alerts;
?

I expect that returning 10000 alerts in a paginated fashion may not be an issue. I was more referring when one tries to retrieve more than that (like beyond >100K alerts) and in that case a paginated search can stress out a cluster. In that case search after or scrolled search are two only two sane ways to get the results out of ES and not using a paginated request at all. A scroll search with or search after search with size of something between 100 and 500 would perform much better.

> That huge performance gain seems surreal, but that is what possible if performance is the goal and the bloat is removed.

I guess it depends a lot on the nature of the queries and documents being percolated.

> we abandoned ES 5.4 product after percolator performance evaluation phase

I hope you reconsider after improvements have been made.

> At a minimum, extra wrapper makes work of query parser slower.

So I was not referring to add this `constant_score` query to your percolator queries, but the `percolate` query in the search api.

> But in ElasticSearch Batch Percolator we noticed 6% increase in speed (from ~5.5 seconds per 1000 percolated alerts down to ~5.2 seconds per 1000 percolated alerts) when we reduced query size by ~11% (from "bool/must" queries to "filtered" queries that ES 5.4 no longer supports)).

Yes, `filtered` has been replaced by `bool` query's `filter` clause. So you should still be able to get this performance improvement?

**Comment by dennisgorelik** [comment]
>> This is argument against implementing standard "from + size" paging for percolator

> Yes.

That means that standard "from + size" paging for percolator feature does not make sense (no sense from business perspective and no sense from performance perspective).
But by merging percolator with main search you added that "from+size" paging feature for percolator, which now requires maintenance (in ElasticSearch codebase) and confuses application developers (like us) who use percolator.

> like beyond >100K alerts

If problems only start at 100K size, then why the maximum page size is only 10K?
Most companies would not even hit that 100K alert results limit (we at postjobfree.com probably will not for few more years). But by giving limited page size you force application developers implement percolator paging up front.
The lack of meaningful size dataset during initial development + forcing to deal with paging issue early + misleading availability of "from+size" paging -- would likely to lead to incorrect custom implementations with a terrible performance results that would start getting noticeable at the time when custom application would scale.

> A scroll search with or search after search with size of something between 100 and 500 would perform much better.

1) Do you mean that that 200 ["search after"](https://www.elastic.co/guide/en/elasticsearch/reference/master/search-request-search-after.html) queries that return 500 alerts each - would run faster than a single query with 100K results?

2) What are the benefits of using "scroll search" over "search after"?
Search after does not need to spend time on percolation of against alerts that are outside of "size" scope. Which means no delay on the first run.
"scroll search" would have that "first run" overhead + the need to maintain that dataset during follow-up queries.

> I hope you reconsider after improvements have been made.

That would be great to have a viable option, but it seems your goal is quite limited: you only plan to bring performance to the levels of ES 1.6 (that requires 3x improvement)
It seems you do not even plan to reach the performance comparable with ElasticSearch batch percolator (which is extra 50x improvement).
It probably would not make sense to sacrifice that 50x performance improvement for the features we do not really need.

> I was not referring to add this constant_score query to your percolator queries, but the percolate query in the search api.

My point here is that query size seems to affect performance. I see how the effect of increased size of "search api" query may be different that the effect of increased "percolator queries" size. But that effect would be negative in both cases, and is likely to be noticeable in both cases.

> Yes, filtered has been replaced by bool query's filter clause. So you should still be able to get this performance improvement?

I think you misunderstood me here.
I meant that "filtered" version of the query (that is no longer supported in ES 5.4) is _faster_ than bool query.
(The difference is relatively small though - only 6% and is probably noticeable because ElasticSearch Batch Percolator core is so fast that queries parsing speed starting to matter).

**Comment by martijnvg** [comment]
Closed in favour of #25445
